File: projects/10-api-gateway/src/balancer.rs
//! V3 — Load balancing across a backend pool.
//!
//! Module: `src/balancer.rs`. A route points at a *pool*; the balancer picks one
//! backend per request. The scaffold gives you the pool structure, the per-backend
//! bookkeeping (in-flight, EWMA latency, circuit breaker), and the policy enum —
//! `pick()` is the `todo!()`. Start with round-robin, then least-connections, then
//! P2C + EWMA, and prove they diverge under skewed load. See SPEC.md V3.

use std::sync::atomic::{AtomicU64, AtomicUsize, Ordering};
use std::sync::Arc;

use serde::Deserialize;

use crate::health::CircuitBreaker;

/// Load-balancing policy over a pool. Deserialized from the route config's `lb`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LbPolicy {
    /// Hand out backends in rotation. Simple, ignores load — the floor.
    #[default]
    RoundRobin,
    /// Route to the backend with the fewest in-flight requests.
    LeastConn,
    /// Power of two choices: sample two at random, pick the less-loaded — the
    /// cheap approximation that avoids round-robin's herd behaviour.
    P2c,
}

/// One upstream backend and its live state.
///
/// The counters and circuit breaker are the *signals* the balancer (V3) and health
/// layer (V4) read. Wiring the accounting into the request path is part of V1/V3.
pub struct Backend {
    /// `host:port` used to build the upstream URI.
    pub addr: String,
    /// Requests currently in flight to this backend (least-conn / P2C signal).
    in_flight: AtomicUsize,
    /// EWMA of recent upstream latency in microseconds (the P2C tie-break signal).
    ewma_micros: AtomicU64,
    /// Per-backend circuit breaker (V4): when open, this backend is skipped.
    pub circuit: CircuitBreaker,
}

impl Backend {
    pub fn new(addr: &str) -> Arc<Self> {
        Arc::new(Self {
            addr: addr.to_string(),
            in_flight: AtomicUsize::new(0),
            ewma_micros: AtomicU64::new(0),
            circuit: CircuitBreaker::new(),
        })
    }

    /// Current in-flight count (least-conn / P2C signal).
    pub fn in_flight(&self) -> usize {
        self.in_flight.load(Ordering::Relaxed)
    }

    /// Increment when a request is dispatched to this backend (V3).
    pub fn incr_in_flight(&self) {
        self.in_flight.fetch_add(1, Ordering::Relaxed);
    }

    /// Decrement when a request to this backend completes (V3).
    pub fn decr_in_flight(&self) {
        self.in_flight.fetch_sub(1, Ordering::Relaxed);
    }

    /// Last recorded EWMA latency in microseconds (P2C tie-break).
    pub fn ewma_micros(&self) -> u64 {
        self.ewma_micros.load(Ordering::Relaxed)
    }

    /// Is this backend eligible to receive a request (circuit not open)? (V4)
    pub fn is_available(&self) -> bool {
        self.circuit.allow()
    }
}

/// The pool + policy for one route's upstream.
pub struct Balancer {
    policy: LbPolicy,
    backends: Vec<Arc<Backend>>,
    /// Rotation cursor for round-robin.
    rr: AtomicUsize,
}

impl Balancer {
    pub fn new(policy: LbPolicy, backends: Vec<Arc<Backend>>) -> Self {
        Self {
            policy,
            backends,
            rr: AtomicUsize::new(0),
        }
    }

    /// The pool (used by the health checker in V4 to probe every backend).
    pub fn backends(&self) -> &[Arc<Backend>] {
        &self.backends
    }

    pub fn policy(&self) -> LbPolicy {
        self.policy
    }

    /// Pick a backend for the next request, or `None` if the whole pool is
    /// unavailable (→ 503).
    ///
    /// TODO(V3): implement per `self.policy` — round-robin (advance `self.rr`),
    /// least-connections (min `in_flight`), then P2C (two random samples, break the
    /// tie on `in_flight`/`ewma_micros`). It must **skip** any backend whose circuit
    /// is open (`is_available() == false`) and must be cheap on the hot path.
    pub fn pick(&self) -> Option<Arc<Backend>> {
        let _ = (&self.rr, &self.backends, self.policy);
        todo!("V3: choose a healthy backend per the LB policy")
    }
}

File: projects/10-api-gateway/src/api_gateway/balancer.py
"""V3 — Load balancing across a backend pool.

Module: `src/api_gateway/balancer.py`. A route points at a *pool*; something has
to pick one backend per request. Round-robin is the floor: it is fair in the
number of requests handed out and completely blind to what happens to them, so
the moment backend #3 starts taking 400 ms instead of 4 ms, round-robin keeps
feeding it exactly one third of your traffic. Naive random is no better — it just
produces unlucky hot spots instead of predictable ones.

The scaffold gives you the pool structure, the per-backend bookkeeping the
policies read (in-flight count, EWMA latency, circuit breaker), and the policy
enum. `Balancer.pick` is yours. Build round-robin, then least-connections, then
P2C + EWMA, and prove with a bench that they diverge under skewed load — the
"prove" half is the actual assignment; all three are ten lines each.

## What "cheap on the hot path" means in Python

The SPEC asks for selection that doesn't serialize every request behind one lock.
In Rust that meant atomics. Here it means something simpler and easy to get wrong
in the other direction: **do not put an `asyncio.Lock` in `pick()`**. This gateway
runs one event loop, so a run of plain attribute reads and writes with no `await`
between them is already indivisible — nothing else can observe a half-updated
cursor. Acquiring a lock would add a scheduler round-trip to every single request
to protect against a race that cannot happen. Keep `pick()` a plain synchronous
function and the problem stays solved.

The corollary is the trap: the *accounting* around a request does span an
`await`. `in_flight` is incremented before the upstream call and decremented
after, and in between the coroutine can be cancelled (client disconnects, deadline
fires) or raise. A decrement that only runs on the success path leaks the counter
upward forever, and since least-connections and P2C read that counter, a leak
quietly and permanently steers traffic *away* from a perfectly healthy backend.
Wrap it in `try` / `finally` — `finally` runs on cancellation too, which is
exactly why it is the right tool here.
"""

from __future__ import annotations

from enum import StrEnum

from .health import CircuitBreaker

__all__ = ["Backend", "Balancer", "LbPolicy"]


class LbPolicy(StrEnum):
    """Load-balancing policy over a pool. Parsed from the route config's `lb`."""

    ROUND_ROBIN = "round_robin"
    """Hand out backends in rotation. Simple, ignores load — the floor."""

    LEAST_CONN = "least_conn"
    """Route to the backend with the fewest in-flight requests."""

    P2C = "p2c"
    """Power of two choices: sample two at random, pick the less-loaded.

    The cheap approximation that beats both. Pure random has no load signal;
    least-connections over a large pool has to look at everything and, worse,
    *every* balancer picks the same "least loaded" backend at the same instant and
    stampedes it. Two random samples cost O(1), and the odds that both are the
    unlucky one are small enough that the herd never forms."""


class Backend:
    """One upstream backend and its live state.

    The counters here are the *signals* the balancer (V3) and the health layer
    (V4) read. None of them update themselves: wiring the accounting into the
    request path is part of V1/V3, and until you do, least-connections and P2C are
    reading zeros and behaving exactly like round-robin — which is a confusing way
    to discover you skipped a step, so do the accounting first.
    """

    __slots__ = ("addr", "circuit", "ewma_seconds", "in_flight")

    def __init__(self, addr: str, circuit: CircuitBreaker | None = None) -> None:
        self.addr = addr
        """`host:port`, used to build the upstream URL."""

        self.in_flight = 0
        """Requests currently in flight to this backend (least-conn / P2C signal).

        Maintained by the proxy path: increment before dispatch, decrement in a
        `finally`. See the module docstring on why `finally` and not "after"."""

        self.ewma_seconds = 0.0
        """Exponentially-weighted moving average of recent upstream latency, in
        seconds (the P2C tie-break signal).

        Seconds, as a float, because that is the unit every asyncio and httpx
        timing API already speaks — the Rust version stored microseconds as an
        integer only because it had to keep it in an atomic."""

        self.circuit = circuit if circuit is not None else CircuitBreaker()
        """Per-backend circuit breaker (V4): when open, this backend is skipped."""

    def is_available(self) -> bool:
        """Is this backend eligible to receive a request (circuit not open)? (V4)"""
        return self.circuit.allow()

    def __repr__(self) -> str:
        return f"Backend({self.addr!r}, in_flight={self.in_flight})"


class Balancer:
    """The pool + policy for one route's upstream."""

    def __init__(self, policy: LbPolicy, backends: list[Backend]) -> None:
        self.policy = policy
        self.backends = backends
        self._cursor = 0
        """Rotation cursor for round-robin. A plain int: see the module docstring."""

    def pick(self) -> Backend | None:
        """Pick a backend for the next request, or `None` if the whole pool is
        unavailable (-> 503, distinct from V2's 404 for *no route at all*).

        TODO(V3): implement per `self.policy` —
          * `ROUND_ROBIN`: advance `self._cursor` modulo the pool size. The subtle
            part is skipping unavailable backends *without* letting the skip
            distort the rotation or spin forever when the pool is entirely down.
          * `LEAST_CONN`: the eligible backend with the smallest `in_flight`.
            `min(eligible, key=...)` is the whole implementation; the interesting
            question is what you break ties on, because a fresh pool is all zeros
            and an arbitrary tie-break silently becomes "always the first one".
          * `P2C`: two distinct random samples from the eligible set
            (`random.sample` gives you distinctness for free), then take the better
            on `in_flight`, breaking ties on `ewma_seconds`. With a pool of one or
            two, P2C degenerates — decide what it does then rather than letting
            `random.sample` raise.

        In every policy: skip any backend whose `is_available()` is `False`, and
        keep the function synchronous and lock-free.
        """
        raise NotImplementedError("V3: choose a healthy backend per the LB policy")
